### ISSUE TITLE
Text inspection acts strange

### ISSUE THREAD

**Issue by kasperosterbye** [guillep/pharo-core#10297]
This bug is really really strange - at least to me. 
Inspecting an instance of Text shows bold, colors, links and all - but not font.

One can do this to reproduce (but it requires visual inspection):
```smalltalk
src := '# First **section**
- foo
- bar'.

txt := (MicRichTextComposer microdownAsRichText: src).
```
Inspecting `txt` gives this:
<img width="83" alt="image" src="https://user-images.githubusercontent.com/17723745/139645539-56e88795-4e04-4c0d-b3ee-92874162cecb.png">

Whereas `txt asMorph` gives:
<img width="145" alt="image" src="https://user-images.githubusercontent.com/17723745/139645616-6b2159b3-e9ac-434d-a632-30c4f8442e53.png">

Installing [*BeautifulComments*](https://github.com/pillar-markup/BeautifulComments), and writing the same markdown in a comment yields this:
<img width="147" alt="image" src="https://user-images.githubusercontent.com/17723745/139646097-6706cf0b-451d-4af0-b0d0-979daeca6a6d.png">

I have been able to track them all back to using `RubScrolledTextMorph` to render Text, but for some odd reason, Font info is omitted in inspect.

**Comment by kasperosterbye** [comment]
Somehow I never discovered it to be the same in Pharo9, but it is.

**Comment by kasperosterbye** [comment]
Some of it seems to be related to a very strange implementation of morph inspection:

```smalltalk
Morph>>inspectionMorph
	<inspectorPresentationOrder: 90 title: 'Morph'>

	^ SpMorphPresenter new
		morph: self imageForm asMorph;
		yourself
```
If one do not do the translation from morph -> image -> morph the ugly bold thingie disappears.
Still no idea why large fonts are not used.

**Comment by kasperosterbye** [comment]
I have traced it down to the following aspects:

Text inspection is done using SpMorphPresenter, and the morph is a RubScrolledTextMorph used to render the text. Spec applies a morphic stylesheet that overwrites some font information in text, accounting for size of the font.

The ugly `txt asMorph` is also related to inspection, in this case morph inspection which is done as:
```smalltalk
Morph>>inspectionMorph
	<inspectorPresentationOrder: 90 title: 'Morph'>

	^ SpMorphPresenter new
		morph: self imageForm asMorph;
		yourself
```

it is the conversion from morph to image and back which blurs it.

The reason it renders correctly in BeautifulComments is because Spec is not involved (do not apply a morphic style to the underlying Rubric Text Area.

**Comment by kasperosterbye** [comment]
I feel like I am stuck between a rock (Rubrics) and a hard place (Spec). I can see two ways to "fix" it, and I do not like any of them.

### At the rubrics side
Add state to `RubAbstractTextArea` which allow me to prevent setting a new font in `textFont:`
```smalltalk
textFont: aFont
        allowFontSetting ifFalse: [^self].
	self addAttribute: (TextFontReference toFont: aFont).
        ...
```

this should then be available as `allowFontSetting` and `disallowFontSetting` on `RubAbstractTextArea` and `RubScrolledTextMorph`

### At the Spec side
Add some state to `SpMorphicLayoutAdapter` to allow a check in `buildMorph: aPresenter constraints: constraints ` to allow a test something like this:

```smalltalk
(aPresenter adapter notNil and: [self allowStyleOverride])
		ifTrue: [ :anAdapter | anAdapter applyStyle ].
```
allowStyleOverride should be true by default (present behaviour), but it should be possible in SpMorphPresenter to do 'blockStyleOverride' or 'allowStyleOverride'

Is there an other way I have not seen?

@estebanlm - could I ask for a bit of your brainpower here?

**Comment by kasperosterbye** [comment]
I finally managed to find the root cause (I believe :-) ). 

There are two types of Font specifications which can be used in `Text`:
<img width="175" alt="image" src="https://user-images.githubusercontent.com/17723745/145161102-7855cc1f-4373-4c54-a321-0bd7d387a9e5.png">

`TextFontChange` is related to font-styles, whereas `TextFontReference` is a specific font. What causes the problem I am struggling with is that the text in question has `TextFontReference` already set, and these should not be overridden by Spec applying a style `(TextFontChange`). In short, a `TextFontChange` should not dominate a `TextFontReference`.

Class `TextFontChange` implements `dominate:` as:
```smalltalk
dominates: other
	^ other isKindOf: TextFontChange
```

This method is used in
```smalltalk
addAttribute: att toArray: others 
	att reset.
	^ Array streamContents:
		[:strm | others do:
			[:other | (att dominates: other) ifFalse: [strm nextPut: other]].
		att set ifTrue: [strm nextPut: att]]
```
If att is a `TextFontChange` and one of the `other` is a `TextFontReference`, it is clear that `TextFontReference` is overridden, which it should not.

**Comment by estebanlm** [comment]
have you tried reproducing it *without* Spec in the middle?
e.g. opening the text with `asMorph openInWindow `  ?
because one thing that may be happening is Spec assigning the default font to the component (which is something that can be worked on)

**Comment by kasperosterbye** [comment]
@estebanlm That produces a good looking result.
Indeed Spec assigns the default font to the component (See heading "At the Spec side" a few comments back). I have not yet a good feeling for any way to solve it. I can very well see cases where "assigning the default font to the component" is the right thing to do (to get consistent interfaces, styles etc). 
One way to "solve" it would be to introduce a flag in SpMorphicLayoutAdapter as described above.

**Comment by estebanlm** [comment]
I will commit a fix.

**Comment by estebanlm** [comment]
> One way to "solve" it would be to introduce a flag in SpMorphicLayoutAdapter as described above.

Nah, we just need to tell morphic presenters to not inherit default font (or any attributes). Since when you are in morphic world, you are technically outside Spec.

**Comment by estebanlm** [comment]
Okay, I submitted the fix and this will be fixed next merge with the image. 
Now, to be honest I think the "good" way to implement this is in fact implementing a `SpMicrodownPresenter` along with its adapter. Using a direct morph in a component that can be so important for Pharo (since we are going to use it everywhere), does not feels right.
 I would gladly guide you to contribute such component.

**Comment by kasperosterbye** [comment]
I would very much like to develop that component. I was actually in the process of doing so, that is how I ran across all these oddities (including Rubrics not being able to handle embedded morphs correctly). 

Do you see a way to do a `SpMicrodownPresenter` without using Rubrics (morph)? What other graphic framework to build on?

I am uncertain which level to attack the issue. At the moment several aspects of the next generation of Microdown rendering will depend on Morph (rendering tables inside a text, as well as multi-column mode). 

Do one need to rewrite the scanners to do this?

**Comment by estebanlm** [comment]
On Dec 8 2021, at 9:14 am, Kasper Østerbye ***@***.***> wrote:
>
>
> I would very much like to develop that component. I was actually in the process of doing so, that is how I ran across all these oddities (including Rubrics not being able to handle embedded morphs correctly).
> Do you see a way to do a SpMicrodownPresenter without using Rubrics (morph)? What other graphic framework to build on?
No, morphic backend will still be Rubric (we do not have anything better than it).

**Comment by kasperosterbye** [comment]
So, a special adapter that will more specifically wrap RubScrolledTextMorph or something. When I have been working on ways to do SpMicrodownPresenter, it has always boiled down to a way to present Text (for instance` aText inspect` also removed fonts as per discussion above - most likely fixed by your commit above),

What kind of overall ideas do you have for making `SpMicrodownPresenter` to kick me in the right direction?

**Comment by estebanlm** [comment]
No specific ideas, tbh. A SpMicrodownPresenter is some kind of a read-only text presenter that shows formatter text. So maybe is a child of `SpAbstractTextPresenter` that is always not editable. An adapter for it should be fairly easy to implement: it receives a string and parses it to a Text to show it using the same old rubric presenter.

**Comment by kasperosterbye** [comment]
OK, then I am on the right track - thanks.

**Comment by kasperosterbye** [comment]
@estebanlm I have now made a triplet of presenter, adapter and morph which allow for read-only view of all to me known attributes of `Text` - it is tentatively named `SpRichTextPresenter`. It has a simple subclass which accepts microdown as string and as document-nodes as input.

I am not sure which package it should be included in - which is my question to you.
* Should it be part of the Spec packages - argument for that is that the three classes are not microdown specific, but basically just fixes a number of issues with rendering `Text` with font attributes (and scaleable images).
* Should it all live in microdown. This feels a bit odd as the `SpRichTextPresenter` should also be used in `Text>>inspectionText` for the newtools inspector as fixes the font error reported way up in this thread.

My suggestion is to 
* let `SpRichTextPresenter` in the `Spec-Core/Widgets-Text` package/tag. The adapter then goes to `Spec2-Adapters-Morphic/Base`. I am not sure where  `SpRichTextMorph` (a subclass of RubScrolledTextMorph which kills font changes and coordinates rescaling of inlined images, but one can view it as part of the adapter, also in `Spec2-Adapters-Morphic/Base`).
* The simple MicrodownPresenter then lives with the MicRichTextComposer in the Microdown packages.

I would like your input before making a PR :-)

**Comment by estebanlm** [comment]
Hi,

I am on holidays right now so I cannot answer completely, sorry :)
What I can tell you is that
a) I would expect a SpRichTextPresenter to be editable somehow, I do not think this is a good name for a readonly component. I would call it SpMicrodownViewPresenter (or similar) instead, which is what the component actually does.
b) It is definitively not part of the Spec2-Core package, since it introduces a dependency on microdown which may or may not be present. I would put this in a separated package (Spec2-Microdown) and in a separated package for the adapter side (Spec2-Adapters-Morphic-Microdown).

To be honest this can live in microdown project correctly, for several reasons:
It is heavily depending on it ;)

I would not replace inspectionText with it! instead, I would add an extra extension inspectionMicrodown, this way we can have the simple inspection and the microdown inspection.

Fibally, I would add this extension to its own package Spec2-Microdown-Inspector-Extensions

I know it sounds verbose, but is the only way to keep it correctly modular :)

cheers!
Esteban

**Comment by kasperosterbye** [comment]
Happy hollidays @estebanlm.

The SpRichTextPresenter reintroduces the functionality to show `Text` which `SpTextPresenter` no longer does. There is **no** dependency from SpRichTextPresenter to microdown in any way, it is purely working on `Text`. The reason the issue is here in the Pharo part is that the current `SpTextPresenter` cannot present `Text` but only `String`. 

My view is that the old TextPresenter did two things:
* Was able to edit Text and String as if they were strings (that is, one could not edit the formatting)
* Was able to show formatted Text (including fonts, colors etc).

The new TextPresenter does one thing:
* Is able to edit Text and String as if they are strings

My component:
* Is able to show formatted Text (including fonts, colors etc)

So, the old TextPresenter is now split in two:
* New TextPresenter which in essence is a StringPresenter
* My component which is in essence a TextViewer (no edit)

Please accept that there is no component in the current Spec which is able to present `Text`. The inspection code in `Text` does not work - that is what this issue is all about!

**Comment by kasperosterbye** [comment]
I see the proposed SpRichTextPresenter (read only) as a parallel to SpImagePresenter (read only). To make an *editor* for Text or Image is a huge task, and not even sure it is necessary.

One way in which one can produce `Text` is using microdown composer, but others exist (the antique help system does it using a simple API), and other might come in the future. 

Once there is a working TextPresenter (presenting font information, not editing font information) it is trivial to make a microdown presenter. A microdown presenter is in essence just:
```smalltalk
WorkingTextPresenter openOn: (MicRichTextComposer compose: (MicParser parse: '# Super nice _heading–'))
``` 
Problem is: There is no `WorkingTextPresenter` in Pharo at the moment.

 I wrote one (currently called `SpRichTextPresenter` as its real name `SpTextPresenter` is used for a thing which should have been called `SpStringPresenter`. 

In the code line above, microdown is used to _produce_ an instance of `Text`, and `SpRichTextPresenter` is not dependent on microdown.

**Comment by estebanlm** [comment]
mmm, I missed a part then.
real problem is the confusion between "Text-the concept" and "Text-the morphic component". 
the text presenter is called such respecting what is a text (or text area) widget in any other existing UI framework in the world :)
So, the `SpTextPresenter` will always be there but is a plain text component. What you are doing is a "morphic-text-viewer" component.
Such component is still not a "SpRichTextComponent" but a "SpMorphicTextViewPresenter" (better names accepted ;). I still want a Rich Text presenter to be editable and not to depending on morphic.

Now, the morphic related/exclusive components needs to be part of `Spec2-Morphic` package and the adapter is part of `Spec2-Adapters-Morphic`. 

I still will not replace inspectionText but add an inspectionMorphicText extention (again, better names accepted). 

EDIT: fix typo

**Comment by estebanlm** [comment]
>. To make an _editor_ for Text or Image is a huge task, and not even sure it is necessary.

Yes, I agree with that, is a huge effort and that's why is not there yet (nor I am asking you to do it!). But I want to reserve the name for a future real rich text editor, and I want to assign proper names to what we have ;)

indeed, this component is in parallel with `SpImagePresenter` (hence the package place I said last message)

**Comment by kasperosterbye** [comment]
The class `Text` is a very old class which has a way to specify text attributes. This predates Morphic (I used it in Smalltalk in early 1990). One has always been able to do:
```smalltalk
('huge' asText addAttribute: 
   (TextFontReference toFont: (LogicalFont 
      familyName: 'Source Sans Pro'
      pointSize: (44)))) inspect
```
and get:
<img width="140" alt="image" src="https://user-images.githubusercontent.com/17723745/146782509-d55b890c-681e-4ca5-8b51-7a8ccfacaf5b.png">

But not in the current version of Pharo (works uptil and including Pharo8, stopped working in 9 and 10 as the only smalltalk not to be able to do this).

So it is not something to do with confusion between "Text-the concept" and "Text-the morphic component", but between "Text-the-concept" and Pharo9 and Pharo10. 

I can follow your argument of reserving `SpTextPresenter` to a future editor (I do not agree though :-) ). The most precise name for the component I have built is perhaps `SpTextInspector` as it fits wits Spec, works with Text, and is read-only (Inspector).  What the adapter and rubric subclass should be named I do not know - they are only used by the `SpTextInspector`. And making it the inspector of Text would return the functionality to be inspected from inspecting an instance of Text.

I did look preliminary at gtk TextTag's which is somewhat similar to TextAttributes, so I believe gtk adapter is possible.
